### ISSUE TITLE
twisted.trial.test.test_assertions.TestAssertionNames.test_failUnless_matches_assert fails with Python trunk@HEAD

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3738]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3738|
|Type|defect|
|Created|2009-03-31 21:07:40Z|
|Branch|https://github.com/twisted/twisted/tree/fail-unless-3738|

```
===============================================================================
[FAIL]: twisted.trial.test.test_assertions.TestAssertionNames.test_failUnless_matches_assert

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/test/test_assertions.py", line 633, in test_failUnless_matches_assert
    dsu(failUnlesses, self._name))
twisted.trial.unittest.FailTest: not equal:
a = [<bound method TestAssertionNames.assertDictContainsSubset of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertDictEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertGreater of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertGreaterEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertIsNone of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertIsNotNone of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertLess of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertLessEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertListEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertMultiLineEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertRaisesRegexp of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertRegexpMatches of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertSameElements of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertSequenceEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertSetEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.assertTupleEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failIf of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnless of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessAlmostEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessAlmostEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessApproximates of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessFailure of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIdentical of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIn of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIsInstance of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessRaises of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessSubstring of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessWarns of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>]
b = [<bound method TestAssertionNames.failIf of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnless of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessAlmostEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessAlmostEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessApproximates of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessEqual of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessFailure of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIdentical of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIn of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessIsInstance of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessRaises of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessSubstring of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>,
 <bound method TestAssertionNames.failUnlessWarns of <twisted.trial.test.test_assertions.TestAssertionNames testMethod=test_failUnless_matches_assert>>]
```

I think this is due to recent unittest.py changes which added numerous new assertion methods, http://svn.python.org/view/python/trunk/Lib/unittest.py?r1=70616&r2=70837

<details><summary>Searchable metadata</summary>

```
trac-id__3738 3738
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__Python_2_7 Python-2.7
branch__branches_fail_unless_3738 branches/fail-unless-3738
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1238533660000000 1238533660000000
changetime__1275784185000000 1275784185000000
version__None None
owner__ 
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [29184]) Branching to 'fail-unless-3738'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank commented|
|-|-|

1. Needs an twisted/topfiles/3738.misc file, probably empty.

2. Please use the newly blessed test method for equality: `self.assertEquals` instead of `self.assertEqual`

3. Can you add a sentence to the test docstring about why the test exists (what mistake it guards against)? I understand what it's asserting, but the purpose is less clear now that the test is less "strict".

Everything else looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [29205]) Remove unused locals and imports; also expand the test method docstring to explain why we're asserting what we are; assertEquals instead of assertEqual, too.

refs [#3738](https://github.com/twisted/twisted/issues/3738)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Fixed those things, I think.  Thanks.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank set owner to @exarkun|
|-|-|

I think this still needs a news file (probably .misc to indicate a minor change), but otherwise it looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [29208]) Merge fail-unless-3738 

Author: therve, exarkun
Reviewer: ivank
Fixes: [#3738](https://github.com/twisted/twisted/issues/3738)

Change a trial test to assert only that the functionality provided by
`failUnless`-prefixed methods is a subset of that provided by `assert`-
prefixed methods.  Previously the test required that these match exactly.

This reflects the preference of `assert`-prefixed methods over `failUnless`-
prefixed methods.  It also fixed the test on Python 2.7, which introduces
several new `assert`-prefixed methods with no corresponding `failUnless`-
prefixed versions.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
